### DISCUSSION TITLE
WebUI and Ollama on Raspi4 problems

### DISCUSSION THREAD

**Discussion by sommersohn** [open-webui/open-webui#5232] [Q&A]
Hi all, 

I am trying to get Ollama and WebUI running on a RaspberryPi4. I don't want to discuss if that makes sense or not. I just want to get it running. I am using docker-compose. Ollama itself seems to work and I can load models and get answers via a python script that accesses port 11434. 
But when I open _http://ip-of-rasppi:8000_ in my browser to access WebUI, I get no answer. 

docker-compose.yaml: 

> version: '3.8'
> 
> services:
>   ollama:
>     image: ollama/ollama
>     container_name: ollama
>     ports:
>       - "8080:8080"
>       - "11434:11434"
>     volumes:
>       - ollama_data:/ollama/data
>     restart: unless-stopped
> 
>   open-webui:
>     image: ghcr.io/open-webui/open-webui:ollama
>     container_name: open-webui
>     ports:
>       - "3000:8080"
>     volumes:
>       - ollama_data:/root/.ollama
>       - open-webui:/app/backend/data
>     restart: always
> 
> volumes:
>   ollama_data:
>   open-webui:

Here is my startup-log: 

> pi@llmraspi:/opt/docker/ollama-web $ sudo docker-compose up
> Starting ollama     ... done
> Starting open-webui ... done
> Attaching to ollama, open-webui
> ollama        | 2024/09/06 16:40:00 routes.go:1125: INFO server config env="map[CUDA_VISIBLE_DEVICES: GPU_DEVICE_ORDINAL: HIP_VISIBLE_DEVICES: HSA_OVERRIDE_GFX_VERSION: OLLAMA_DEBUG:false OLLAMA_FLASH_ATTENTION:false OLLAMA_HOST:http://0.0.0.0:11434 OLLAMA_INTEL_GPU:false OLLAMA_KEEP_ALIVE:5m0s OLLAMA_LLM_LIBRARY: OLLAMA_MAX_LOADED_MODELS:0 OLLAMA_MAX_QUEUE:512 OLLAMA_MODELS:/root/.ollama/models OLLAMA_NOHISTORY:false OLLAMA_NOPRUNE:false OLLAMA_NUM_PARALLEL:0 OLLAMA_ORIGINS:[http://localhost https://localhost http://localhost:* https://localhost:* http://127.0.0.1 https://127.0.0.1 http://127.0.0.1:* https://127.0.0.1:* http://0.0.0.0 https://0.0.0.0 http://0.0.0.0:* https://0.0.0.0:* app://* file://* tauri://*] OLLAMA_RUNNERS_DIR: OLLAMA_SCHED_SPREAD:false OLLAMA_TMPDIR: ROCR_VISIBLE_DEVICES:]"
> open-webui    | Loading WEBUI_SECRET_KEY from file, not provided as an environment variable.
> open-webui    | Loading WEBUI_SECRET_KEY from .webui_secret_key
> open-webui    | USE_OLLAMA is set to true, starting ollama serve.
> ollama        | time=2024-09-06T16:40:00.664Z level=INFO source=images.go:753 msg="total blobs: 6"
> ollama        | time=2024-09-06T16:40:00.665Z level=INFO source=images.go:760 msg="total unused blobs removed: 0"
> ollama        | time=2024-09-06T16:40:00.666Z level=INFO source=routes.go:1172 msg="Listening on [::]:11434 (version 0.3.9)"
> ollama        | time=2024-09-06T16:40:00.669Z level=INFO source=payload.go:30 msg="extracting embedded files" dir=/tmp/ollama794250401/runners
> open-webui    | 2024/09/06 16:40:00 routes.go:1125: INFO server config env="map[CUDA_VISIBLE_DEVICES: GPU_DEVICE_ORDINAL: HIP_VISIBLE_DEVICES: HSA_OVERRIDE_GFX_VERSION: OLLAMA_DEBUG:false OLLAMA_FLASH_ATTENTION:false OLLAMA_HOST:http://127.0.0.1:11434 OLLAMA_INTEL_GPU:false OLLAMA_KEEP_ALIVE:5m0s OLLAMA_LLM_LIBRARY: OLLAMA_MAX_LOADED_MODELS:0 OLLAMA_MAX_QUEUE:512 OLLAMA_MODELS:/root/.ollama/models OLLAMA_NOHISTORY:false OLLAMA_NOPRUNE:false OLLAMA_NUM_PARALLEL:0 OLLAMA_ORIGINS:[http://localhost https://localhost http://localhost:* https://localhost:* http://127.0.0.1 https://127.0.0.1 http://127.0.0.1:* https://127.0.0.1:* http://0.0.0.0 https://0.0.0.0 http://0.0.0.0:* https://0.0.0.0:* app://* file://* tauri://*] OLLAMA_RUNNERS_DIR: OLLAMA_SCHED_SPREAD:false OLLAMA_TMPDIR: ROCR_VISIBLE_DEVICES:]"
> open-webui    | time=2024-09-06T16:40:00.814Z level=INFO source=images.go:782 msg="total blobs: 0"
> open-webui    | time=2024-09-06T16:40:00.814Z level=INFO source=images.go:790 msg="total unused blobs removed: 0"
> open-webui    | time=2024-09-06T16:40:00.815Z level=INFO source=routes.go:1172 msg="Listening on 127.0.0.1:11434 (version 0.3.6)"
> open-webui    | time=2024-09-06T16:40:01.230Z level=WARN source=assets.go:89 msg="process still running, skipping" pid=11 path=/tmp/ollama4076775309/ollama.pid
> open-webui    | time=2024-09-06T16:40:01.232Z level=WARN source=assets.go:89 msg="process still running, skipping" pid=13 path=/tmp/ollama4178440021/ollama.pid
> open-webui    | time=2024-09-06T16:40:01.236Z level=INFO source=payload.go:30 msg="extracting embedded files" dir=/tmp/ollama733125929/runners

**Comment by silentoplayz** [comment]
Your Open WebUI instance would be accessible on Port `3000` according to this, not Port `8000`.
![image](https://github.com/user-attachments/assets/3646a1ef-6d57-49cf-8fe5-aead8e3eb25c)

**Comment by justinh-rahb** [comment]
No idea why you're opening port 8080 on the Ollama container, there's nothing listening on that. Also you've pulled the `:ollama` image of WebUI which has Ollama included as well, so you're running two instances of it.

**Comment by sommersohn** [comment]
Yes. Correct. It must be port 3000 to open in Webbrowser. But,  if I try to open the WebUI on port 3000, it also doesn't open it: 

> pi@llmraspi:~ $ sudo docker ps -a
> CONTAINER ID   IMAGE                                  COMMAND               CREATED      STATUS                          PORTS                                                                                      NAMES
> 18ab964f329c   ghcr.io/open-webui/open-webui:ollama   "bash start.sh"       3 days ago   Up 7 hours (health: starting)   0.0.0.0:3000->8080/tcp, :::3000->8080/tcp                                                  open-webui
> 91d263ef1d16   ollama/ollama                          "/bin/ollama serve"   3 days ago   Up 21 hours                     0.0.0.0:8080->8080/tcp, :::8080->8080/tcp, 0.0.0.0:11434->11434/tcp, :::11434->11434/tcp   ollama
> pi@llmraspi:~ $ curl http://10.0.0.249:11434
> Ollama is running
> pi@llmraspi:~ $ curl http://10.0.0.249:3000
> curl: (7) Failed to connect to 10.0.0.249 port 3000 after 0 ms: Couldn't connect to server

**Reply by justinh-rahb** [reply]
Unsurprising, as explained above, you did it wrong.

**Comment by cyberb** [comment]
if Ollama is working fine on port 11434 but you cant reach WebUI in the browser, its almost certainly a Docker networking issue. by default Open WebUI tries to connect to Ollama at localhost which inside the container means the container itself, not your Pi.

try running with --network=host flag so the container shares your Pis network stack. or use --add-host=host.docker.internal:host-gateway and set OLLAMA_BASE_URL to http://host.docker.internal:11434

also the RPi4 is pretty tight on resources for Open WebUI. make sure you have at least 4GB RAM, the WebUI itself eats a decent chunk even before loading models.

if the whole docker networking thing is more trouble than its worth, theres Syncloud (https://github.com/syncloud/platform) which has Open WebUI as a one click install for raspberry pi with the networking pre-configured. i built it so full disclosure im biased, but it handles exactly this port/networking stuff automatically
